Sources/Build/Sanitizers.swift

```diff
@@ -0,0 +1,100 @@
+
+ Copyright (c) 2018 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+import Basic
+import Utility
+
+/// A set of enabled runtime sanitizers.
+public struct EnabledSanitizers {
+    public enum Sanitizer: String {
+        case address
+        case thread
+    }
+
+    /// A set of enabled sanitizers.
+    private var sanitizers: Set<Sanitizer> = []
+
+    public init() {}
+    public init(from list: [Sanitizer]) {
+        self.sanitizers = Set(list)
+    }
+
+    /// Return an established short name for a sanitizer, e.g. "asan".
+    private func shortname(_ sanitizer: Sanitizer) -> String {
+        switch sanitizer {
+        case .address: return "asan"
+        case .thread: return "tsan"
+        }
+    }
+
+    public func runtimeEnvironment(toolchain: Toolchain) -> [String: String] {
+      #if os(macOS)
+        return macosRuntimeEnvironment(toolchain: toolchain)
+      #else
+        return Process.env
+      #endif
+    }
+
+    public func macosRuntimeEnvironment(toolchain: Toolchain) -> [String: String] {
+        var env = Process.env
+        var insertLibs: [String] = []
+
+        if let existingLibs = env["DYLD_INSERT_LIBRARIES"] {
+            insertLibs = [existingLibs]
+        }
+
+        for sanitizer in sanitizers {
+            let libPath = sanitizerLibrary(clangPath: toolchain.clangCompiler, for: sanitizer)
+            insertLibs.append(libPath.asString)
+        }
+
+        env["DYLD_INSERT_LIBRARIES"] = insertLibs.joined(separator: ":")
+
+        if Process.verbose {
+            env["DYLD_PRINT_LIBRARIES"] = "1"
+        }
+
+        return env
+    }
+
+    /// Sanitization flags for the Swift compiler
+    public func compileSwiftFlags() -> [String] {
+        return sanitizers.map { "-sanitize=\($0.rawValue)" }
+    }
+
+    /// Sanitization flags for the Swift linker are the same
+    /// as for the compiler.
+    public func linkSwiftFlags() -> [String] {
+        return compileSwiftFlags()
+    }
+
+    /// The sanitizer library is part of clang, but it is possible
+    /// to fetch a proper version of it by going through the `swift` symlink.
+    private func sanitizerLibrary(clangPath clang: AbsolutePath, for sanitizer: Sanitizer) -> AbsolutePath {
+        return clang.appending(RelativePath("../../lib/swift/clang/lib/\(osName())/libclang_rt.\(shortname(sanitizer))_osx_dynamic.dylib"))
+    }
+
+    /// Operating system short name for the filesystem path.
+    private func osName() -> String {
+        switch Triple.hostTriple.os {
+        case .darwin: return "darwin"
+        case .macOS: return "darwin"  // Still darwin for our purposes
+        case .linux: return "linux"
+        }
+    }
+
+}
+
+/// Adding a protocol to help with command line option parsing.
```

Comment:
nit: don't think we require this comment.

Reply:
This comment is the only thing that visually ties the extension with the concept of the command line option parsing. Leaving it.

Reply:
`SourceKit` will automatically inherit comments from the protocol.

Reply:
That's just one way to read the sources. The other is print them or read in vi.

Reply:
Hah, okay. Can you change the comment to something like:

```
/// StringEnumArgument Conformation to help with command line parsing.
```

Reply:
I meant `conformation` and not `Conformation`.